### ISSUE TITLE
Synchrony through cleanup

### ISSUE THREAD

**Issue by tcstewar** [ctn-waterloo/modelling_ideas#18]
What happens if I take the output of a cleanup and subtract it from the input to the cleanup?  

It seems to me that it should start up some oscillatory behaviour where if the original input originally had DOG+CAT, it'd end up oscillating between DOG and CAT.  The phases would be random, but have the same period.  If there was a bit of mutual inhibition between the nodes in the cleanup, that'd force the phases to be out of sync.

Note: I seem to remember results from the cockroach olfactory system that shows this sort of thing for novel odours (known odours cause oscillatory patterns, even in combination, but novel odours cause a constant pattern).

**Comment by tcstewar** [comment]
Seems to work:

```python
import nengo
import nengo.spa as spa

D = 32

model = spa.SPA()
with model:
    model.vision = spa.Buffer(D)
    model.recog = spa.Buffer(D)
    
    vocab = model.get_output_vocab('recog')
    vocab.parse('DOG+CAT')
    model.cleanup = spa.AssociativeMemory(vocab) 
    
    model.cortical=spa.Cortical(
        spa.Actions(
            'recog = vision',
            'cleanup = recog',
            'recog = -cleanup',
            ))
```

If you feed in DOG to vision, then recog starts oscillating with DOG.  If you feed in DOG+CAT, you get DOG and CAT oscillating out of phase.  If you feed in DOG+CAT+RAT, DOG and CAT oscillate while RAT stays somewhat constant.

**Comment by tcstewar** [comment]
Even better:

```python
           'recog = -10*cleanup',
```

**Comment by tcstewar** [comment]
Here's a version that shows that this approach can accentuate differences between inputs:

```python
import nengo
import nengo.spa as spa

D = 32

model = spa.SPA()
with model:
    model.vision = spa.Buffer(D)
    model.recog = spa.Buffer(D)
    
    vocab = model.get_output_vocab('recog')
    S1 = vocab.parse('S1')
    S2 = vocab.parse('S1+0.4*R2')
    S2.normalize()
    vocab.add('S2', S2)
    model.cleanup = spa.AssociativeMemory(vocab) 

    model.cortical=spa.Cortical(
        spa.Actions(
            'recog = vision',
            'cleanup = recog',
            'recog = -10*cleanup',
            ))
```

The inputs S1 and S2 are very similar, but you get very different firing patterns in the cleanup part for these two inputs.  This could be useful for categorization.

**Comment by Seanny123** [comment]
What paper did you see that described these firing patterns? [Couldn't find it myself](https://cogsci.stackexchange.com/q/18261/4397).

**Comment by tcstewar** [comment]
> What paper did you see that described these firing patterns? Couldn't find it myself.

Good question...  this one comes close but isn't quite what I remember...
http://www.sciencedirect.com/science/article/pii/S1084952106000437?via%3Dihub

**Comment by tcstewar** [comment]
This one is a bit closer:

http://brain.mpg.de/fileadmin/user_upload/Documents/Papers/Papers_Laurent/StopferLaurent_Nature99.pdf

". Interpolating a novel odour within a
sequence of stimulus with a familiar odour did not affect the
response to the familiar odour (not shown). Exposure to an
odour, however, increased the probability of oscillatory synchronization
upon the first presentation of a chemically related odour"

**Comment by tcstewar** [comment]
> The inputs S1 and S2 are very similar, but you get very different firing patterns in the cleanup part for these two inputs. This could be useful for categorization.

I'll also note that both of the papers I linked above indicate that this oscillatory behaviour helps for distinguishing highly similar inputs...  :)

**Comment by Seanny123** [comment]
Turns out the classifications works so well directly decoding from the oscillating ensemble, that you don't even need a temporal network. The results are super clean and scale REALLY well for:

- short presentation time (50ms)
- large amounts of classes (128) 
- small dimensional input (16)
- high similarity (around 0.85) [note that I am still trying to figure out [how to generate vectors with a given similarity better](https://forum.nengo.ai/t/generating-highly-similar-spa-vectors/404?u=seanny123)]

Check out [my two notebooks](https://github.com/Seanny123/oscillation_decode) for further details.

**Comment by Seanny123** [comment]
Upon further review, it doesn't actually work that well and performs almost identically to just decoding from the ensemble, even with really high similarity. So I'll proceed with testing with the rolling window network, which from what I understand will require D dimensions, being fed into D rolling window networks.

**Comment by Seanny123** [comment]
So, to apply decoders from a rolling network, you would have to concatenate the state output somehow, because they can only handle a single dimensional input.

As Aaron pointed out, the only reason the oscillator would work is if the oscillation contains some sort of information in it's history that's valuable. This would be equivalent to just using a delay node to delay the output of the oscillator and concatenate it with the output, which I tried and actually hurt the classification.

Additionally, the ensemble and the oscillator almost work identically, even with super high similarities like 0.9999.

In conclusion, as I discussed with Aaron, because we don't really know what the oscillator is doing, nor do we know how to make a better oscillator that would oscillate in a more ideal manner, we can't really expect the oscillation to improve classification. So, I'm going to go ahead and close this issue until someone figures out how to use a delay in classification for better accuracy other than just integrating over time.

**Comment by tcstewar** [comment]
I'm going to reopen the issue, since the original point was just to produce synchrony, rather than to use the synchrony to do classification....  It is definitely weird to me and a fascinating result that it looks like this synchrony doesn't help at all for classification, and anyone who goes and explores this idea further should look into that, but it is separate from the original point of the issue.   

Thank you all for the work looking into this!  I certainly learned lots from talking about it with you...  :)
